engine/src/main/java/io/zeebe/engine/nwe/container/CallActivityProcessor.java

```diff
@@ -0,0 +1,199 @@
+  }
+
+  @Override
+  public void onCompleted(final ExecutableCallActivity element, final BpmnElementContext context) {
+    stateTransitionBehavior.takeOutgoingSequenceFlows(element, context);
+    stateBehavior.consumeToken(context);
+    stateBehavior.removeElementInstance(context);
+  }
+
+  @Override
+  public void onTerminating(
+      final ExecutableCallActivity element, final BpmnElementContext context) {
+    stateTransitionBehavior.terminateChildProcessInstance(context);
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+  }
+
+  @Override
+  public void onTerminated(final ExecutableCallActivity element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.publishTriggeredBoundaryEvent(context);
+    incidentBehavior.resolveIncidents(context);
+    stateTransitionBehavior.onElementTerminated(element, context);
+    stateBehavior.consumeToken(context);
+    stateBehavior.removeElementInstance(context);
+  }
+
+  @Override
+  public void onEventOccurred(
+      final ExecutableCallActivity element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.triggerBoundaryEvent(element, context);
+  }
+
+  @Override
+  public void onChildCompleted(
+      final ExecutableCallActivity element,
+      final BpmnElementContext callActivityContext,
+      final BpmnElementContext childContext) {
+    final var currentState = callActivityContext.getIntent();
+    switch (currentState) {
+      case ELEMENT_ACTIVATED:
+        stateTransitionBehavior.transitionToCompleting(callActivityContext);
+        break;
+      case ELEMENT_TERMINATING:
+        // the call activity is already terminating, it doesn't matter that the child completed
+        // do nothing
+        break;
+      default:
+        final var message = String.format(UNABLE_TO_COMPLETE_FROM_STATE_MESSAGE, currentState);
+        throw new BpmnProcessingException(callActivityContext, message);
+    }
+  }
+
+  @Override
+  public void onChildTerminated(
+      final ExecutableCallActivity element,
+      final BpmnElementContext callActivityContext,
+      final BpmnElementContext childContext) {
+    final var currentState = callActivityContext.getIntent();
+    if (currentState == WorkflowInstanceIntent.ELEMENT_TERMINATING) {
+      stateTransitionBehavior.transitionToTerminated(callActivityContext);
+    } else {
+      final var message = String.format(UNABLE_TO_TERMINATE_FROM_STATE_MESSAGE, currentState);
+      throw new BpmnProcessingException(callActivityContext, message);
+    }
+  }
+
+  private Either<Failure, DirectBuffer> evaluateProcessId(
+      final BpmnElementContext context, final ExecutableCallActivity element) {
+    final var processIdExpression = element.getCalledElementProcessId();
+    final var scopeKey = context.getElementInstanceKey();
+    return expressionProcessor
+        .evaluateStringExpression(processIdExpression, scopeKey)
+        .map(BufferUtil::wrapString);
```

Comment:
To avoid the creation of a new buffer, the method `evaluateStringExpression()` could return the buffer directly. There is a method in the expression processor that is not migrated yet.